File: Escenario11_Actividad5_IA.cpp
// ESCENARIO 11: MSA e Incertidumbre de la Medición
// Análisis del Sistema de Medición usando Naive Bayes
// Archivo: Escenario11_Actividad5_IA.cpp

#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>
#include <random>
#include <iomanip>

using namespace std;

class NaiveBayes {
private:
    map<int, vector<vector<double>>> clases;
    map<int, vector<pair<double, double>>> estadisticas; // media, desviacion
    
    double calcularMedia(const vector<double>& datos) {
        double suma = 0;
        for (double val : datos) {
            suma += val;
        }
        return suma / datos.size();
    }
    
    double calcularDesviacion(const vector<double>& datos) {
        double media = calcularMedia(datos);
        double sumaCuadrados = 0;
        for (double val : datos) {
            sumaCuadrados += pow(val - media, 2);
        }
        double varianza = sumaCuadrados / datos.size();
        return varianza > 0 ? sqrt(varianza) : 0.0001;
    }
    
    double probabilidadGaussiana(double x, double media, double desviacion) {
        double exponente = exp(-pow(x - media, 2) / (2 * pow(desviacion, 2)));
        return (1.0 / (sqrt(2 * M_PI) * desviacion)) * exponente;
    }
    
public:
    void entrenar(const vector<vector<double>>& X, const vector<int>& y) {
        // Agrupar datos por clase
        for (size_t i = 0; i < X.size(); i++) {
            int clase = y[i];
            clases[clase].push_back(X[i]);
        }
        
        // Calcular estadisticas por clase
        for (auto& par : clases) {
            int clase = par.first;
            vector<vector<double>>& valores = par.second;
            
            size_t numCaracteristicas = valores[0].size();
            for (size_t j = 0; j < numCaracteristicas; j++) {
                vector<double> columna;
                for (const auto& fila : valores) {
                    columna.push_back(fila[j]);
                }
                
                double media = calcularMedia(columna);
                double desviacion = calcularDesviacion(columna);
                estadisticas[clase].push_back({media, desviacion});
            }
        }
    }
    
    map<int, double> predecirProba(const vector<double>& x) {
        map<int, double> probabilidades;
        
        for (auto& par : estadisticas) {
            int clase = par.first;
            vector<pair<double, double>>& stats = par.second;
            double prob = 1.0;
            
            for (size_t i = 0; i < x.size(); i++) {
                double media = stats[i].first;
                double desviacion = stats[i].second;
                prob *= probabilidadGaussiana(x[i], media, desviacion);
            }
            probabilidades[clase] = prob;
        }
        
        // Normalizar probabilidades
        double total = 0;
        for (auto& par : probabilidades) {
            total += par.second;
        }
        if (total > 0) {
            for (auto& par : probabilidades) {
                par.second /= total;
            }
        }
        
        return probabilidades;
    }
    
    int predecir(const vector<double>& x) {
        map<int, double> probabilidades = predecirProba(x);
        int mejorClase = -1;
        double mejorProb = -1;
        
        for (auto& par : probabilidades) {
            if (par.second > mejorProb) {
                mejorProb = par.second;
                mejorClase = par.first;
            }
        }
        
        return mejorClase;
    }
};

vector<vector<double>> generarDataset(int n, mt19937& gen) {
    vector<vector<double>> datos(n, vector<double>(6));
    
    normal_distribution<double> distRepet(5.0, 8.0);
    normal_distribution<double> distReprod(4.0, 7.0);
    normal_distribution<double> distNdc(3.0, 3.0);
    normal_distribution<double> distLin(2.0, 4.0);
    normal_distribution<double> distEst(3.0, 5.0);
    
    for (int i = 0; i < n; i++) {
        // 1. Repetibilidad (%)
        datos[i][0] = abs(distRepet(gen));
        
        // 2. Reproducibilidad (%)
        datos[i][1] = abs(distReprod(gen));
        
        // 3. GRR Total (%)
        datos[i][2] = sqrt(pow(datos[i][0], 2) + pow(datos[i][1], 2));
        
        // 4. ndc (Número de categorías distintas)
        datos[i][3] = abs(distNdc(gen));
        
        // 5. Linearidad (%)
        datos[i][4] = abs(distLin(gen));
        
        // 6. Estabilidad (%)
        datos[i][5] = abs(distEst(gen));
    }
    
    return datos;
}

vector<int> generarEtiquetas(const vector<vector<double>>& X) {
    vector<int> etiquetas(X.size());
    
    for (size_t i = 0; i < X.size(); i++) {
        double repetibilidad = X[i][0];
        double reproducibilidad = X[i][1];
        double grr = X[i][2];
        double ndc = X[i][3];
        double linearidad = X[i][4];
        double estabilidad = X[i][5];
        
        // Criterios de clasificación según estándares MSA
        // Clase 2: ACEPTABLE
        if (grr < 10.0 && ndc >= 5 && linearidad < 5.0 && estabilidad < 10.0) {
            etiquetas[i] = 2;
        }
        // Clase 1: MARGINAL
        else if (grr < 30.0 && ndc >= 3 && linearidad < 10.0 && estabilidad < 20.0) {
            etiquetas[i] = 1;
        }
        // Clase 0: INACEPTABLE
        else {
            etiquetas[i] = 0;
        }
    }
    
    return etiquetas;
}

void dividirDataset(const vector<vector<double>>& X, const vector<int>& y,
                   vector<vector<double>>& X_train, vector<int>& y_train,
                   vector<vector<double>>& X_test, vector<int>& y_test,
                   double testSize, mt19937& gen) {
    int n = X.size();
    int nTest = (int)(n * testSize);
    int nTrain = n - nTest;
    
    vector<int> indices(n);
    for (int i = 0; i < n; i++) {
        indices[i] = i;
    }
    shuffle(indices.begin(), indices.end(), gen);
    
    X_train.resize(nTrain);
    y_train.resize(nTrain);
    X_test.resize(nTest);
    y_test.resize(nTest);
    
    for (int i = 0; i < nTrain; i++) {
        int idx = indices[i];
        X_train[i] = X[idx];
        y_train[i] = y[idx];
    }
    
    for (int i = 0; i < nTest; i++) {
        int idx = indices[i + nTrain];
        X_test[i] = X[idx];
        y_test[i] = y[idx];
    }
}

double calcularAccuracy(const vector<int>& y_true, const vector<int>& y_pred) {
    int correctos = 0;
    for (size_t i = 0; i < y_true.size(); i++) {
        if (y_true[i] == y_pred[i]) {
            correctos++;
        }
    }
    return (double)correctos / y_true.size();
}

void imprimirMatrizConfusion(const vector<int>& y_true, const vector<int>& y_pred, int numClases) {
    vector<vector<int>> matriz(numClases, vector<int>(numClases, 0));
    
    for (size_t i = 0; i < y_true.size(); i++) {
        matriz[y_true[i]][y_pred[i]]++;
    }
    
    cout << "\n=== MATRIZ DE CONFUSION ===" << endl;
    cout << "              ";
    for (int i = 0; i < numClases; i++) {
        cout << "Pred_" << i << "  ";
    }
    cout << endl;
    
    string nombres[] = {"Inaceptable", "Marginal   ", "Aceptable  "};
    for (int i = 0; i < numClases; i++) {
        cout << "Real_" << i << " (" << nombres[i] << "): ";
        for (int j = 0; j < numClases; j++) {
            cout << setw(6) << matriz[i][j] << "  ";
        }
        cout << endl;
    }
}

map<int, map<string, double>> calcularMetricas(const vector<int>& y_true, 
                                                const vector<int>& y_pred, 
                                                int numClases) {
    map<int, map<string, double>> metricas;
    
    for (int clase = 0; clase < numClases; clase++) {
        int tp = 0, fp = 0, fn = 0, tn = 0;
        
        for (size_t i = 0; i < y_true.size(); i++) {
            if (y_true[i] == clase && y_pred[i] == clase) tp++;
            else if (y_true[i] != clase && y_pred[i] == clase) fp++;
            else if (y_true[i] == clase && y_pred[i] != clase) fn++;
            else tn++;
        }
        
        double precision = (tp + fp > 0) ? (double)tp / (tp + fp) : 0;
        double recall = (tp + fn > 0) ? (double)tp / (tp + fn) : 0;
        double f1 = (precision + recall > 0) ? 2 * precision * recall / (precision + recall) : 0;
        
        metricas[clase]["precision"] = precision;
        metricas[clase]["recall"] = recall;
        metricas[clase]["f1"] = f1;
    }
    
    return metricas;
}

double calcularCostoMejora(const vector<int>& y_true, const vector<int>& y_pred) {
    // Costos asociados a sistemas de medición
    const double COSTO_CALIBRACION = 2000.0;
    const double COSTO_CAPACITACION = 1500.0;
    const double COSTO_EQUIPO_NUEVO = 15000.0;
    const double COSTO_PRODUCTO_DEFECTUOSO = 500.0;
    
    double costoTotal = 0;
    int sistemasInaceptables = 0;
    int sistemasMarginales = 0;
    
    for (size_t i = 0; i < y_true.size(); i++) {
        // Contar sistemas por categoría real
        if (y_true[i] == 0) {
            sistemasInaceptables++;
            costoTotal += COSTO_EQUIPO_NUEVO;
        } else if (y_true[i] == 1) {
            sistemasMarginales++;
            costoTotal += COSTO_CALIBRACION + COSTO_CAPACITACION;
        }
        
        // Penalización por clasificación incorrecta
        if (y_true[i] != y_pred[i]) {
            if (y_true[i] == 0 && y_pred[i] > 0) {
                costoTotal += COSTO_PRODUCTO_DEFECTUOSO * 10;
            } else if (y_true[i] > 0 && y_pred[i] == 0) {
                costoTotal += COSTO_CALIBRACION;
            }
        }
    }
    
    return costoTotal;
}

int main() {
    cout << "========================================" << endl;
    cout << "ESCENARIO 11: MSA E INCERTIDUMBRE DE MEDICION" << endl;
    cout << "Analisis del Sistema de Medicion (MSA)" << endl;
    cout << "========================================\n" << endl;
    
    mt19937 gen(42);
    
    // Generar dataset
    cout << "Generando dataset de estudios MSA..." << endl;
    vector<vector<double>> X = generarDataset(400, gen);
    vector<int> y = generarEtiquetas(X);
    
    // Contar clases
    map<int, int> conteoClases;
    for (int clase : y) {
        conteoClases[clase]++;
    }
    
    cout << "\nDistribucion de sistemas de medicion:" << endl;
    cout << "- Inaceptable (0): " << conteoClases[0] << endl;
    cout << "- Marginal (1): " << conteoClases[1] << endl;
    cout << "- Aceptable (2): " << conteoClases[2] << endl;
    
    // Dividir dataset
    vector<vector<double>> X_train, X_test;
    vector<int> y_train, y_test;
    dividirDataset(X, y, X_train, y_train, X_test, y_test, 0.2, gen);
    
    cout << "\nConjunto de entrenamiento: " << X_train.size() << " estudios" << endl;
    cout << "Conjunto de prueba: " << X_test.size() << " estudios" << endl;
    
    // Entrenar modelo
    cout << "\nEntrenando clasificador Naive Bayes..." << endl;
    NaiveBayes modelo;
    modelo.entrenar(X_train, y_train);
    
    // Realizar predicciones
    cout << "Realizando predicciones..." << endl;
    vector<int> y_pred(X_test.size());
    for (size_t i = 0; i < X_test.size(); i++) {
        y_pred[i] = modelo.predecir(X_test[i]);
    }
    
    // Calcular accuracy
    double accuracy = calcularAccuracy(y_test, y_pred);
    cout << "\n=== RESULTADOS ===" << endl;
    cout << fixed << setprecision(2);
    cout << "Accuracy: " << accuracy * 100 << "%" << endl;
    
    // Matriz de confusión
    imprimirMatrizConfusion(y_test, y_pred, 3);
    
    // Métricas por clase
    map<int, map<string, double>> metricas = calcularMetricas(y_test, y_pred, 3);
    cout << "\n=== METRICAS POR CLASE ===" << endl;
    string nombresClases[] = {"Inaceptable", "Marginal", "Aceptable"};
    
    for (int i = 0; i < 3; i++) {
        cout << "\nClase " << i << " (" << nombresClases[i] << "):" << endl;
        cout << "  Precision: " << metricas[i]["precision"] * 100 << "%" << endl;
        cout << "  Recall: " << metricas[i]["recall"] * 100 << "%" << endl;
        cout << "  F1-Score: " << metricas[i]["f1"] * 100 << "%" << endl;
    }
    
    // Análisis de costos
    double costoMejora = calcularCostoMejora(y_test, y_pred);
    cout << "\n=== ANALISIS ECONOMICO ===" << endl;
    cout << "Costo estimado de mejora/reemplazo: $" << costoMejora << " USD" << endl;
    cout << "Costo promedio por sistema: $" << costoMejora / y_test.size() << " USD" << endl;
    
    // Ejemplo de predicción
    cout << "\n=== EJEMPLO DE PREDICCION ===" << endl;
    vector<double> ejemplo = {8.5, 6.2, 10.6, 4.2, 3.8, 8.5};
    cout << "Sistema de medicion evaluado:" << endl;
    cout << "  Repetibilidad (EV): " << ejemplo[0] << "%" << endl;
    cout << "  Reproducibilidad (AV): " << ejemplo[1] << "%" << endl;
    cout << "  GRR Total: " << ejemplo[2] << "%" << endl;
    cout << "  ndc: " << ejemplo[3] << endl;
    cout << "  Linearidad: " << ejemplo[4] << "%" << endl;
    cout << "  Estabilidad: " << ejemplo[5] << "%" << endl;
    
    int prediccion = modelo.predecir(ejemplo);
    map<int, double> probabilidades = modelo.predecirProba(ejemplo);
    
    cout << "\nPrediccion: " << nombresClases[prediccion] << endl;
    cout << "Probabilidades:" << endl;
    for (int i = 0; i < 3; i++) {
        cout << "  " << nombresClases[i] << ": " << probabilidades[i] * 100 << "%" << endl;
    }
    
    // Interpretación según estándares MSA
    cout << "\n=== INTERPRETACION SEGUN MSA ===" << endl;
    if (prediccion == 2) {
        cout << "✓ Sistema ACEPTABLE: Puede usarse sin restricciones" << endl;
        cout << "  - GRR < 10%: Excelente variacion del sistema" << endl;
        cout << "  - ndc >= 5: Resolucion adecuada" << endl;
    } else if (prediccion == 1) {
        cout << "⚠ Sistema MARGINAL: Aceptable con reservas" << endl;
        cout << "  - GRR entre 10-30%: Requiere mejoras" << endl;
        cout << "  - Considerar: calibracion y capacitacion" << endl;
    } else {
        cout << "✗ Sistema INACEPTABLE: NO usar para mediciones criticas" << endl;
        cout << "  - GRR > 30%: Variacion excesiva" << endl;
        cout << "  - Accion requerida: Reemplazar equipo o proceso" << endl;
    }
    
    // Preguntas de reflexión
    cout << "\n=== PREGUNTAS DE REFLEXION ===" << endl;
    cout << "1. ¿Por que un GRR < 10% se considera aceptable segun AIAG?" << endl;
    cout << "2. ¿Cual es la diferencia entre repetibilidad y reproducibilidad?" << endl;
    cout << "3. ¿Que significa que ndc = 2 en un sistema de medicion?" << endl;
    cout << "4. ¿Como afecta un sistema marginal a las decisiones de calidad?" << endl;
    cout << "5. ¿Es etico usar IA para clasificar sistemas de medicion automaticamente?" << endl;
    cout << "6. ¿Que impacto tiene un falso negativo (sistema malo clasificado como bueno)?" << endl;
    
    return 0;
}

File: Escenario11_Actividad5_IA.java
// ESCENARIO 11: MSA e Incertidumbre de la Medición
// Análisis del Sistema de Medición usando Naive Bayes
// Archivo: Escenario11_Actividad5_IA.java

import java.util.*;

class NaiveBayes {
    private Map<Integer, List<double[]>> clases;
    private Map<Integer, List<double[]>> estadisticas;
    
    public NaiveBayes() {
        clases = new HashMap<>();
        estadisticas = new HashMap<>();
    }
    
    private double calcularMedia(List<Double> datos) {
        double suma = 0;
        for (double val : datos) {
            suma += val;
        }
        return suma / datos.size();
    }
    
    private double calcularDesviacion(List<Double> datos) {
        double media = calcularMedia(datos);
        double sumaCuadrados = 0;
        for (double val : datos) {
            sumaCuadrados += Math.pow(val - media, 2);
        }
        double varianza = sumaCuadrados / datos.size();
        return varianza > 0 ? Math.sqrt(varianza) : 0.0001;
    }
    
    private double probabilidadGaussiana(double x, double media, double desviacion) {
        double exponente = Math.exp(-Math.pow(x - media, 2) / (2 * Math.pow(desviacion, 2)));
        return (1 / (Math.sqrt(2 * Math.PI) * desviacion)) * exponente;
    }
    
    public void entrenar(double[][] X, int[] y) {
        // Agrupar datos por clase
        for (int i = 0; i < X.length; i++) {
            int clase = y[i];
            if (!clases.containsKey(clase)) {
                clases.put(clase, new ArrayList<>());
            }
            clases.get(clase).add(X[i]);
        }
        
        // Calcular estadisticas por clase
        for (Map.Entry<Integer, List<double[]>> entry : clases.entrySet()) {
            int clase = entry.getKey();
            List<double[]> valores = entry.getValue();
            List<double[]> stats = new ArrayList<>();
            
            int numCaracteristicas = valores.get(0).length;
            for (int j = 0; j < numCaracteristicas; j++) {
                List<Double> columna = new ArrayList<>();
                for (double[] fila : valores) {
                    columna.add(fila[j]);
                }
                double media = calcularMedia(columna);
                double desviacion = calcularDesviacion(columna);
                stats.add(new double[]{media, desviacion});
            }
            estadisticas.put(clase, stats);
        }
    }
    
    public Map<Integer, Double> predecirProba(double[] x) {
        Map<Integer, Double> probabilidades = new HashMap<>();
        
        for (Map.Entry<Integer, List<double[]>> entry : estadisticas.entrySet()) {
            int clase = entry.getKey();
            List<double[]> stats = entry.getValue();
            double prob = 1.0;
            
            for (int i = 0; i < x.length; i++) {
                double media = stats.get(i)[0];
                double desviacion = stats.get(i)[1];
                prob *= probabilidadGaussiana(x[i], media, desviacion);
            }
            probabilidades.put(clase, prob);
        }
        
        // Normalizar probabilidades
        double total = 0;
        for (double prob : probabilidades.values()) {
            total += prob;
        }
        if (total > 0) {
            for (int clase : probabilidades.keySet()) {
                probabilidades.put(clase, probabilidades.get(clase) / total);
            }
        }
        
        return probabilidades;
    }
    
    public int predecir(double[] x) {
        Map<Integer, Double> probabilidades = predecirProba(x);
        int mejorClase = -1;
        double mejorProb = -1;
        
        for (Map.Entry<Integer, Double> entry : probabilidades.entrySet()) {
            if (entry.getValue() > mejorProb) {
                mejorProb = entry.getValue();
                mejorClase = entry.getKey();
            }
        }
        
        return mejorClase;
    }
}

public class Escenario11_Actividad5_IA {
    
    private static double[][] generarDataset(int n, Random rand) {
        double[][] datos = new double[n][6];
        
        for (int i = 0; i < n; i++) {
            // Características del sistema de medición
            // 1. Repetibilidad (%) - Variación del equipo
            datos[i][0] = 5.0 + rand.nextGaussian() * 8.0;
            
            // 2. Reproducibilidad (%) - Variación entre operadores
            datos[i][1] = 4.0 + rand.nextGaussian() * 7.0;
            
            // 3. GRR Total (%) - Gage R&R
            datos[i][2] = Math.sqrt(Math.pow(datos[i][0], 2) + Math.pow(datos[i][1], 2));
            
            // 4. ndc (Número de categorías distintas)
            datos[i][3] = 3.0 + rand.nextGaussian() * 3.0;
            
            // 5. Linearidad (%)
            datos[i][4] = 2.0 + rand.nextGaussian() * 4.0;
            
            // 6. Estabilidad (%)
            datos[i][5] = 3.0 + rand.nextGaussian() * 5.0;
            
            // Asegurar valores positivos
            for (int j = 0; j < 6; j++) {
                if (datos[i][j] < 0) datos[i][j] = Math.abs(datos[i][j]);
            }
        }
        
        return datos;
    }
    
    private static int[] generarEtiquetas(double[][] X) {
        int[] etiquetas = new int[X.length];
        
        for (int i = 0; i < X.length; i++) {
            double repetibilidad = X[i][0];
            double reproducibilidad = X[i][1];
            double grr = X[i][2];
            double ndc = X[i][3];
            double linearidad = X[i][4];
            double estabilidad = X[i][5];
            
            // Criterios de clasificación según estándares MSA
            // Clase 2: ACEPTABLE (Sistema de medición confiable)
            if (grr < 10.0 && ndc >= 5 && linearidad < 5.0 && estabilidad < 10.0) {
                etiquetas[i] = 2;
            }
            // Clase 1: MARGINAL (Sistema aceptable con reservas)
            else if (grr < 30.0 && ndc >= 3 && linearidad < 10.0 && estabilidad < 20.0) {
                etiquetas[i] = 1;
            }
            // Clase 0: INACEPTABLE (Sistema de medición no confiable)
            else {
                etiquetas[i] = 0;
            }
        }
        
        return etiquetas;
    }
    
    private static void dividirDataset(double[][] X, int[] y, 
                                      double[][] X_train, int[] y_train,
                                      double[][] X_test, int[] y_test,
                                      double testSize, Random rand) {
        int n = X.length;
        int nTest = (int)(n * testSize);
        int nTrain = n - nTest;
        
        List<Integer> indices = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            indices.add(i);
        }
        Collections.shuffle(indices, rand);
        
        for (int i = 0; i < nTrain; i++) {
            int idx = indices.get(i);
            X_train[i] = X[idx].clone();
            y_train[i] = y[idx];
        }
        
        for (int i = 0; i < nTest; i++) {
            int idx = indices.get(i + nTrain);
            X_test[i] = X[idx].clone();
            y_test[i] = y[idx];
        }
    }
    
    private static double calcularAccuracy(int[] y_true, int[] y_pred) {
        int correctos = 0;
        for (int i = 0; i < y_true.length; i++) {
            if (y_true[i] == y_pred[i]) {
                correctos++;
            }
        }
        return (double)correctos / y_true.length;
    }
    
    private static void imprimirMatrizConfusion(int[] y_true, int[] y_pred, int numClases) {
        int[][] matriz = new int[numClases][numClases];
        
        for (int i = 0; i < y_true.length; i++) {
            matriz[y_true[i]][y_pred[i]]++;
        }
        
        System.out.println("\n=== MATRIZ DE CONFUSION ===");
        System.out.print("              ");
        for (int i = 0; i < numClases; i++) {
            System.out.printf("Pred_%d  ", i);
        }
        System.out.println();
        
        String[] nombres = {"Inaceptable", "Marginal   ", "Aceptable  "};
        for (int i = 0; i < numClases; i++) {
            System.out.printf("Real_%d (%s): ", i, nombres[i]);
            for (int j = 0; j < numClases; j++) {
                System.out.printf("%6d  ", matriz[i][j]);
            }
            System.out.println();
        }
    }
    
    private static Map<Integer, Map<String, Double>> calcularMetricas(int[] y_true, int[] y_pred, int numClases) {
        Map<Integer, Map<String, Double>> metricas = new HashMap<>();
        
        for (int clase = 0; clase < numClases; clase++) {
            int tp = 0, fp = 0, fn = 0, tn = 0;
            
            for (int i = 0; i < y_true.length; i++) {
                if (y_true[i] == clase && y_pred[i] == clase) tp++;
                else if (y_true[i] != clase && y_pred[i] == clase) fp++;
                else if (y_true[i] == clase && y_pred[i] != clase) fn++;
                else tn++;
            }
            
            double precision = (tp + fp > 0) ? (double)tp / (tp + fp) : 0;
            double recall = (tp + fn > 0) ? (double)tp / (tp + fn) : 0;
            double f1 = (precision + recall > 0) ? 2 * precision * recall / (precision + recall) : 0;
            
            Map<String, Double> metricasClase = new HashMap<>();
            metricasClase.put("precision", precision);
            metricasClase.put("recall", recall);
            metricasClase.put("f1", f1);
            
            metricas.put(clase, metricasClase);
        }
        
        return metricas;
    }
    
    private static double calcularCostoMejora(int[] y_true, int[] y_pred) {
        // Costos asociados a sistemas de medición
        final double COSTO_CALIBRACION = 2000.0;        // Costo de calibración profesional
        final double COSTO_CAPACITACION = 1500.0;       // Capacitación de operadores
        final double COSTO_EQUIPO_NUEVO = 15000.0;      // Reemplazo de equipo
        final double COSTO_PRODUCTO_DEFECTUOSO = 500.0; // Producto mal clasificado
        
        double costoTotal = 0;
        int sistemasInaceptables = 0;
        int sistemasMarginales = 0;
        
        for (int i = 0; i < y_true.length; i++) {
            // Contar sistemas por categoría real
            if (y_true[i] == 0) {
                sistemasInaceptables++;
                costoTotal += COSTO_EQUIPO_NUEVO; // Requiere reemplazo
            } else if (y_true[i] == 1) {
                sistemasMarginales++;
                costoTotal += COSTO_CALIBRACION + COSTO_CAPACITACION; // Requiere mejora
            }
            
            // Penalización por clasificación incorrecta
            if (y_true[i] != y_pred[i]) {
                if (y_true[i] == 0 && y_pred[i] > 0) {
                    // Sistema inaceptable clasificado como bueno - MUY GRAVE
                    costoTotal += COSTO_PRODUCTO_DEFECTUOSO * 10;
                } else if (y_true[i] > 0 && y_pred[i] == 0) {
                    // Sistema bueno clasificado como malo - DESPERDICIO
                    costoTotal += COSTO_CALIBRACION;
                }
            }
        }
        
        return costoTotal;
    }
    
    public static void main(String[] args) {
        System.out.println("========================================");
        System.out.println("ESCENARIO 11: MSA E INCERTIDUMBRE DE MEDICION");
        System.out.println("Analisis del Sistema de Medicion (MSA)");
        System.out.println("========================================\n");
        
        Random rand = new Random(42);
        
        // Generar dataset
        System.out.println("Generando dataset de estudios MSA...");
        double[][] X = generarDataset(400, rand);
        int[] y = generarEtiquetas(X);
        
        // Contar clases
        Map<Integer, Integer> conteoClases = new HashMap<>();
        for (int clase : y) {
            conteoClases.put(clase, conteoClases.getOrDefault(clase, 0) + 1);
        }
        
        System.out.println("\nDistribucion de sistemas de medicion:");
        System.out.println("- Inaceptable (0): " + conteoClases.getOrDefault(0, 0));
        System.out.println("- Marginal (1): " + conteoClases.getOrDefault(1, 0));
        System.out.println("- Aceptable (2): " + conteoClases.getOrDefault(2, 0));
        
        // Dividir dataset
        int nTest = (int)(X.length * 0.2);
        int nTrain = X.length - nTest;
        
        double[][] X_train = new double[nTrain][6];
        int[] y_train = new int[nTrain];
        double[][] X_test = new double[nTest][6];
        int[] y_test = new int[nTest];
        
        dividirDataset(X, y, X_train, y_train, X_test, y_test, 0.2, rand);
        
        System.out.println("\nConjunto de entrenamiento: " + X_train.length + " estudios");
        System.out.println("Conjunto de prueba: " + X_test.length + " estudios");
        
        // Entrenar modelo
        System.out.println("\nEntrenando clasificador Naive Bayes...");
        NaiveBayes modelo = new NaiveBayes();
        modelo.entrenar(X_train, y_train);
        
        // Realizar predicciones
        System.out.println("Realizando predicciones...");
        int[] y_pred = new int[X_test.length];
        for (int i = 0; i < X_test.length; i++) {
            y_pred[i] = modelo.predecir(X_test[i]);
        }
        
        // Calcular accuracy
        double accuracy = calcularAccuracy(y_test, y_pred);
        System.out.println("\n=== RESULTADOS ===");
        System.out.printf("Accuracy: %.2f%%\n", accuracy * 100);
        
        // Matriz de confusión
        imprimirMatrizConfusion(y_test, y_pred, 3);
        
        // Métricas por clase
        Map<Integer, Map<String, Double>> metricas = calcularMetricas(y_test, y_pred, 3);
        System.out.println("\n=== METRICAS POR CLASE ===");
        String[] nombresClases = {"Inaceptable", "Marginal", "Aceptable"};
        
        for (int i = 0; i < 3; i++) {
            Map<String, Double> m = metricas.get(i);
            System.out.printf("\nClase %d (%s):\n", i, nombresClases[i]);
            System.out.printf("  Precision: %.2f%%\n", m.get("precision") * 100);
            System.out.printf("  Recall: %.2f%%\n", m.get("recall") * 100);
            System.out.printf("  F1-Score: %.2f%%\n", m.get("f1") * 100);
        }
        
        // Análisis de costos
        double costoMejora = calcularCostoMejora(y_test, y_pred);
        System.out.println("\n=== ANALISIS ECONOMICO ===");
        System.out.printf("Costo estimado de mejora/reemplazo: $%.2f USD\n", costoMejora);
        System.out.printf("Costo promedio por sistema: $%.2f USD\n", costoMejora / y_test.length);
        
        // Ejemplo de predicción
        System.out.println("\n=== EJEMPLO DE PREDICCION ===");
        double[] ejemplo = {8.5, 6.2, 10.6, 4.2, 3.8, 8.5};
        System.out.println("Sistema de medicion evaluado:");
        System.out.printf("  Repetibilidad (EV): %.2f%%\n", ejemplo[0]);
        System.out.printf("  Reproducibilidad (AV): %.2f%%\n", ejemplo[1]);
        System.out.printf("  GRR Total: %.2f%%\n", ejemplo[2]);
        System.out.printf("  ndc: %.1f\n", ejemplo[3]);
        System.out.printf("  Linearidad: %.2f%%\n", ejemplo[4]);
        System.out.printf("  Estabilidad: %.2f%%\n", ejemplo[5]);
        
        int prediccion = modelo.predecir(ejemplo);
        Map<Integer, Double> probabilidades = modelo.predecirProba(ejemplo);
        
        System.out.println("\nPrediccion: " + nombresClases[prediccion]);
        System.out.println("Probabilidades:");
        for (int i = 0; i < 3; i++) {
            System.out.printf("  %s: %.2f%%\n", nombresClases[i], probabilidades.get(i) * 100);
        }
        
        // Interpretación según estándares MSA
        System.out.println("\n=== INTERPRETACION SEGUN MSA ===");
        if (prediccion == 2) {
            System.out.println("✓ Sistema ACEPTABLE: Puede usarse sin restricciones");
            System.out.println("  - GRR < 10%: Excelente variacion del sistema");
            System.out.println("  - ndc >= 5: Resolucion adecuada");
        } else if (prediccion == 1) {
            System.out.println("⚠ Sistema MARGINAL: Aceptable con reservas");
            System.out.println("  - GRR entre 10-30%: Requiere mejoras");
            System.out.println("  - Considerar: calibracion y capacitacion");
        } else {
            System.out.println("✗ Sistema INACEPTABLE: NO usar para mediciones criticas");
            System.out.println("  - GRR > 30%: Variacion excesiva");
            System.out.println("  - Accion requerida: Reemplazar equipo o proceso");
        }
        
        // Preguntas de reflexión
        System.out.println("\n=== PREGUNTAS DE REFLEXION ===");
        System.out.println("1. ¿Por que un GRR < 10% se considera aceptable segun AIAG?");
        System.out.println("2. ¿Cual es la diferencia entre repetibilidad y reproducibilidad?");
        System.out.println("3. ¿Que significa que ndc = 2 en un sistema de medicion?");
        System.out.println("4. ¿Como afecta un sistema marginal a las decisiones de calidad?");
        System.out.println("5. ¿Es etico usar IA para clasificar sistemas de medicion automaticamente?");
        System.out.println("6. ¿Que impacto tiene un falso negativo (sistema malo clasificado como bueno)?");
    }
}